### PR TITLE
 Use std::vectory instead of malloc in solver allocation

### LINKED ISSUE #1304: Use new instead of malloc in solver allocation
src/solver/simulation/sim_alloc_probleme_hebdo.cpp

### PR DESCRIPTION
**Pull request by payetvin**
close #1304

### CODE REVIEW

**Review by flomnes — CHANGES_REQUESTED** [review]
Don't use `free` in this case, use `delete` or `delete[]`

File: src/solver/simulation/sim_alloc_probleme_hebdo.cpp [hunk 1]
```diff
@@ -46,70 +46,64 @@ void SIM_AllocationProblemeHebdo(PROBLEME_HEBDO& problem, uint NombreDePasDeTemp
     const uint linkCount = study.runtime->interconnectionsCount();
     const uint shortTermStorageCount = study.runtime->shortTermStorageCount;
 
-    problem.DefaillanceNegativeUtiliserPMinThermique = (bool*)MemAlloc(nbPays * sizeof(char));
-    problem.DefaillanceNegativeUtiliserHydro = (bool*)MemAlloc(nbPays * sizeof(char));
-    problem.DefaillanceNegativeUtiliserConsoAbattue = (bool*)MemAlloc(nbPays * sizeof(char));
+    problem.DefaillanceNegativeUtiliserPMinThermique = std::vector<bool>(nbPays);
```
Comment (flomnes) [inline src/solver/simulation/sim_alloc_probleme_hebdo.cpp]:
```suggestion
    problem.DefaillanceNegativeUtiliserPMinThermique.assign(nbPays, false);
```

**Review by flomnes — CHANGES_REQUESTED** [review]

File: src/solver/simulation/sim_alloc_probleme_hebdo.cpp [hunk 2]
```diff
@@ -46,307 +46,270 @@ void SIM_AllocationProblemeHebdo(PROBLEME_HEBDO& problem, uint NombreDePasDeTemp
...
-      = (SOLDE_MOYEN_DES_ECHANGES**)MemAlloc(NombreDePasDeTemps * sizeof(void*));
+    problem.DefaillanceNegativeUtiliserPMinThermique.assign(nbPays, false);
+    problem.DefaillanceNegativeUtiliserHydro.assign(nbPays, false);
+    problem.DefaillanceNegativeUtiliserConsoAbattue.assign(nbPays, false);
+
+    problem.CoefficientEcretementPMaxHydraulique.assign(nbPays, 0.);
+
+    problem.BruitSurCoutHydraulique.assign(nbPays, std::vector<double>(8784));
+
+    problem.NomsDesPays = std::vector<const char*>(nbPays);
```
Comment (guilpier-code) [inline src/solver/simulation/sim_alloc_probleme_hebdo.cpp]: About : `problem.NomsDesPays = std::vector<const char*>(nbPays);`
Why not : 
```C++
problem.NomsDesPays.resize(nbPays);
```
? 
Same remark for other assignment like this

**Review by guilpier-code — DISMISSED** [review]

File: src/solver/simulation/sim_alloc_probleme_hebdo.cpp (hunk 2 above)
Comment (flomnes) [reply src/solver/simulation/sim_alloc_probleme_hebdo.cpp]: I agree, please use `assign` for consistency and to make sure memory is initialized.

File: src/solver/simulation/sim_structure_probleme_economique.h [hunk 3]
```diff
@@ -561,58 +548,56 @@ struct PROBLEME_HEBDO
     bool LeProblemeADejaEteInstancie = false;
     bool firstWeekOfSimulation = false;
 
-    CORRESPONDANCES_DES_VARIABLES** CorrespondanceVarNativesVarOptim = nullptr;
-    CORRESPONDANCES_DES_CONTRAINTES** CorrespondanceCntNativesCntOptim = nullptr;
-    CORRESPONDANCES_DES_CONTRAINTES_JOURNALIERES** CorrespondanceCntNativesCntOptimJournalieres = nullptr;
-    CORRESPONDANCES_DES_CONTRAINTES_HEBDOMADAIRES** CorrespondanceCntNativesCntOptimHebdomadaires = nullptr;
+    std::vector<CORRESPONDANCES_DES_VARIABLES*> CorrespondanceVarNativesVarOptim;
```
Comment (flomnes) [inline src/solver/simulation/sim_structure_probleme_economique.h:551]: Please leave a TODO note here, and also an explanation on what's not working

File: src/solver/simulation/sim_alloc_probleme_hebdo.cpp [hunk 4]
```diff
@@ -46,295 +46,253 @@ void SIM_AllocationProblemeHebdo(PROBLEME_HEBDO& problem, unsigned NombreDePasDe
...
-    problem.PaliersThermiquesDuPays = new PALIERS_THERMIQUES*[nbPays];
-    problem.CaracteristiquesHydrauliques = new ENERGIES_ET_PUISSANCES_HYDRAULIQUES*[nbPays];
-    problem.previousSimulationFinalLevel = new double[nbPays];
+    problem.DefaillanceNegativeUtiliserPMinThermique.assign(nbPays, false);
+    problem.DefaillanceNegativeUtiliserHydro.assign(nbPays, false);
+    problem.DefaillanceNegativeUtiliserConsoAbattue.assign(nbPays, false);
+
+    problem.CoefficientEcretementPMaxHydraulique.assign(nbPays, 0.);
+
+    problem.BruitSurCoutHydraulique.assign(nbPays, std::vector<double>(8784));
```
Comment (flomnes) [inline src/solver/simulation/sim_alloc_probleme_hebdo.cpp:55]: Use named constants, not hard-coded numbers.

File: src/solver/simulation/common-hydro-levels.cpp [hunk 5]
```diff
@@ -62,15 +63,13 @@ void computingHydroLevels(const Data::AreaList& areas,
...
+        double nivInit = problem.CaracteristiquesHydrauliques[index].NiveauInitialReservoir;
         double* niv = weeklyResults.niveauxHoraires;
 
         double* ovf = weeklyResults.debordementsHoraires;
 
-        auto& computeLvlObj = problem.computeLvl_object;
-
-        computeLvlObj.setParameters(
-          nivInit, inflows, ovf, turb, pumpingRatio, pump, reservoirCapacity);
+        auto computeLvlObj = computeTimeStepLevel(
```
Comment (flomnes) [inline src/solver/simulation/common-hydro-levels.cpp]:
```suggestion
        computeTimeStepLevel computeLvlObj(
```

**Review by flomnes — CHANGES_REQUESTED** [review]
LGTM. Only minor changes requested.

**Comment by flomnes** [comment]
Please also leave a TODO note for structures used to store results.

File: src/solver/simulation/solver.hxx [hunk 6]
```diff
@@ -146,7 +146,7 @@ private:
             double* randomReservoirLevel = nullptr;
 
             if (hydroHotStart && firstSetParallelWithAPerformedYearWasRun)
-                randomReservoirLevel = state[numSpace].problemeHebdo->previousYearFinalLevels;
+                randomReservoirLevel = &state[numSpace].problemeHebdo->previousYearFinalLevels[0];
```
Comment (flomnes) [inline src/solver/simulation/solver.hxx]:
```suggestion
                randomReservoirLevel = state[numSpace].problemeHebdo->previousYearFinalLevels.data();
```

File: src/solver/simulation/sim_alloc_probleme_hebdo.cpp [hunk 7]
```diff
@@ -580,259 +535,82 @@ void SIM_DesallocationProblemeHebdo(PROBLEME_HEBDO& problem)
...
+                      .PuissanceDisponibleEtCout[j]
+                      ->NombreMaxDeGroupesEnMarcheDuPalierThermique);
+            MemFree(problem.PaliersThermiquesDuPays[k]
+                      .PuissanceDisponibleEtCout[j]
+                      ->NombreMinDeGroupesEnMarcheDuPalierThermique);
+            MemFree(problem.PaliersThermiquesDuPays[k]
+                      .PuissanceDisponibleEtCout[j]
+                      ->CoutHoraireDuPalierThermiqueUp);
+            MemFree(problem.PaliersThermiquesDuPays[k]
+                      .PuissanceDisponibleEtCout[j]
```
Comment (flomnes) [inline src/solver/simulation/sim_alloc_probleme_hebdo.cpp]: Be careful, you're mixing `new` (C++) and `free` (C)

**Review by flomnes — CHANGES_REQUESTED** [review]

File: src/solver/simulation/sim_alloc_probleme_hebdo.cpp (hunk 7 above)
Comment (payetvin) [reply src/solver/simulation/sim_alloc_probleme_hebdo.cpp]: Will be deleted when replaced with vectors

**Review by flomnes — APPROVED** [review]

**Pull request merged**